compiler/src/iree/compiler/InputConversion/TMTensor/ConvertTMTensorToLinalgExt.cpp

```diff
@@ -69,6 +75,85 @@ struct ScatterOpConversion
 };
 }  // namespace
 
+static Value collapseBatches(PatternRewriter &rewriter, Location loc,
+                             Value val) {
+  auto valSizes = val.getType().cast<RankedTensorType>().getShape();
+  int64_t newBatch =
+      std::accumulate(valSizes.begin(), valSizes.end() - 2, 1,
+                      [](int64_t x, int64_t y) { return x * y; });
+  Type elementType = val.getType().cast<RankedTensorType>().getElementType();
+  SmallVector<int64_t> newSizes{newBatch};
+  newSizes.append(valSizes.end() - 2, valSizes.end());
+  Type newType = RankedTensorType::get(newSizes, elementType);
+
+  auto rank = valSizes.size();
+  SmallVector<int64_t> collapsed;
+  for (auto i = 0; i < rank - 2; i++) collapsed.push_back(i);
+
+  SmallVector<ReassociationIndices> reassociation(3);
+  reassociation[0].append(collapsed);
+  reassociation[1].push_back(rank - 2);
+  reassociation[2].push_back(rank - 1);
+
+  return rewriter
+      .create<tensor::CollapseShapeOp>(loc, newType, val, reassociation)
+      .getResult();
+}
+static Value expandBatches(PatternRewriter &rewriter, Location loc,
+                           SmallVector<int64_t> batchSizes, Value val) {
+  auto valSizes = val.getType().cast<RankedTensorType>().getShape();
+  Type elementType = val.getType().cast<RankedTensorType>().getElementType();
+  SmallVector<int64_t> newSizes(batchSizes);
+  newSizes.append(valSizes.end() - 2, valSizes.end());
+  auto rank = newSizes.size();
+  Type newType = RankedTensorType::get(newSizes, elementType);
+
+  SmallVector<ReassociationIndices> reassociation(3);
+  for (auto i = 0; i < batchSizes.size(); i++) reassociation[0].push_back(i);
+  reassociation[1].push_back(rank - 2);
+  reassociation[2].push_back(rank - 1);
+
+  return rewriter
+      .create<tensor::ExpandShapeOp>(loc, newType, val, reassociation)
+      .getResult();
+}
```

Comment:
This lowering detail (collapsing/expanding batches) feels like a frontend detail that could have been handled in advance to allow parity between the TMTensor and LinalgExt versions of attention.

At least that matches my understanding of the two dialects but would be helpful if someone who knows the intended relationship of the two dialects well could clarify. I wouldn't think changing the upstream op at this point would be worth it though.

Reply:
I thought it made more sense for the tm_tensor op to correspond to the torch op rather than the backend op to try to keep tm_tensor backend-agnostic. We can change it if that makes more sense.

Reply:
With some offline discussion, the plan is to update the TMTensor op definition in torch-mlir to match the linalg_ext op in a separate revision. For now this is just copying the tm_tensor.attention op definition.